src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -415,15 +502,231 @@ private void CheckDisposedOrStarted()
+                    {
+                        retVal.DangerousAddRef(ref gotReference);
+                        rawHandle = retVal.DangerousGetHandle();
+                    }
+                    rawHandle = Interop.libcurl.curl_slist_append(rawHandle, header);
+                    retVal.SetHandle(rawHandle);
+                }
+            }
+            finally
+            {
+                if (gotReference)
+                {
+                    retVal.DangerousRelease();
+                }
+            }
+
+            if (!retVal.IsInvalid)
+            {
+                try
+                {
+                    retVal.DangerousAddRef(ref gotReference);
+                    SetCurlOption(handle, CURLoption.CURLOPT_HTTPHEADER, retVal.DangerousGetHandle());
+                }
+                finally
+                {
+                    if (gotReference)
+                    {
+                        retVal.DangerousRelease();
+                    }
+                }
+            }
+
+            return retVal;
+        }
+
+        private void AddEasyHandle(RequestCompletionSource state)
+        {
+            bool gotReference = false;
+            SafeCurlHandle requestHandle = state.RequestHandle;
+            try
+            {
+                requestHandle.DangerousAddRef(ref gotReference);
+                int result = Interop.libcurl.curl_multi_add_handle(_multiHandle, requestHandle);
+                if (result != CURLcode.CURLE_OK)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result, true));
+                }
+                state.SessionHandle = _multiHandle;
+            }
+            catch (Exception)
+            {
+                if (gotReference)
+                {
+                    requestHandle.DangerousRelease();
+                }
+                throw;
+            }
+        }
+
+        private static void RemoveEasyHandle(SafeCurlMultiHandle multiHandle, GCHandle stateHandle, bool onMultiStack)
         {
-            return new Exception(GetCurlErrorString(code));
+            RequestCompletionSource state = (RequestCompletionSource)stateHandle.Target;
+            SafeCurlHandle requestHandle = state.RequestHandle;
+
+            if (onMultiStack)
+            {
+                Interop.libcurl.curl_multi_remove_handle(multiHandle, requestHandle);
+                state.SessionHandle = null;
+                requestHandle.DangerousRelease();
+            }
+
+            if (!state.RequestHeaderHandle.IsInvalid)
+            {
+                SafeCurlSlistHandle headerHandle = state.RequestHeaderHandle;
+                SafeCurlSlistHandle.DisposeAndClearHandle(ref headerHandle);
+                state.RequestHeaderHandle.SetHandleAsInvalid();
+            }
+
+            SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
+            state.RequestHandle.SetHandleAsInvalid();
```

Comment:
Why are we sometimes referring to it as state.RequestHandle and sometimes as requestHandle?  That confuses me.  I'm also not clear why we're calling SetHandleAsInvalid on it when we just disposed it... just for good measure?